client/txn.go

```diff
@@ -384,47 +378,40 @@ func (txn *Txn) exec(retryable func(txn *Txn) error) error {
 // always commit or clean-up explicitly even when that may not be
 // required (or even erroneous).
 func (txn *Txn) send(calls ...proto.Call) error {
+	if txn.Proto.Status != proto.PENDING {
+		return util.Errorf("attempting to use %s transaction", txn.Proto.Status)
+	}
+
 	lastIndex := len(calls) - 1
 	if lastIndex < 0 {
 		return nil
 	}
 
+	lastReq := calls[lastIndex].Args
 	// haveTxnWrite tracks intention to write. This is in contrast to
 	// txn.Proto.Writing, which is set by the coordinator when the first
 	// intent has been created, and which lives for the life of the
 	// transaction.
-	var haveTxnWrite bool
-	var haveEndTxn bool
+	haveTxnWrite := proto.IsTransactionWrite(lastReq)
 
-	for i, call := range calls {
+	for _, call := range calls[:lastIndex] {
```

Comment:
>  "We only care about haveTxnWrite because it allows us to chop off an unnecessary EndTransactionRequest. Therefore the write status of the last call never matters, only whether it is an EndTransactionRequest."

The write status of the last call does matter, in the case that it is not an `EndTrasactionRequest`. Consider the case of a single `Put` call as the simplest example.

> Whichever level is constructing the EndTransaction calls.

In practice, the `EndTransaction` call comes from one of three places:
- sql/server.go calling `Commit()` 
- `exec()` calling `commit()`
- "user" code calling `CommitInBatch()` (currently only used in TxnCoordSender).

I don't understand why we'd want to push any logic out to those places, but maybe I'm just being thick.